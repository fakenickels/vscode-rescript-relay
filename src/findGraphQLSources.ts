import { getLocator } from "locate-character";
import { GraphQLSource, GraphQLSourceFromTag } from "./extensionTypes";

/**
 * A helper for extracting GraphQL operations from source via a regexp.
 * It assumes that the only thing the regexp matches is the actual content,
 * so if that's not true for your regexp you probably shouldn't use this
 * directly.
 */
export let makeExtractTagsFromSource = (
  regexp: RegExp
): ((text: string) => Array<GraphQLSourceFromTag>) => (
  text: string
): Array<GraphQLSourceFromTag> => {
  const locator = getLocator(text);
  const sources: Array<GraphQLSourceFromTag> = [];
  let result;
  while ((result = regexp.exec(text)) !== null) {
    let start = locator(result.index);
    let end = locator(result.index + result[0].length);

    sources.push({
      type: "TAG",
      content: result[0],
      start: {
        line: start.line,
        character: start.column,
      },
      end: {
        line: end.line,
        character: end.column,
      },
    });
  }

  return sources;
};

export const jsGraphQLTagsRegexp = new RegExp(
  /(?<=(graphql|gql|graphql\.experimental)`)[.\s\S]+?(?=`)/g
);
export const rescriptFileFilterRegexp = new RegExp(/(\%relay\()/g);
export const rescriptGraphQLTagsRegexp = new RegExp(
  /(?<=\%relay\([\s]*`)[\s\S.]+?(?=`[\s]*\))/g
);

const extractGraphQLSourceFromJs = makeExtractTagsFromSource(
  jsGraphQLTagsRegexp
);

export const extractGraphQLSourceFromReScript = makeExtractTagsFromSource(
  rescriptGraphQLTagsRegexp
);

export function extractGraphQLSources(
  languageId: string,
  document: string
): GraphQLSource[] | null {
  switch (languageId) {
    case "graphql":
      return [
        {
          type: "FULL_DOCUMENT",
          content: document,
        },
      ];
    case "javascript":
    case "javascriptreact":
    case "typescript":
    case "typescriptreact":
    case "vue":
      return extractGraphQLSourceFromJs(document);
    case "rescript":
      return extractGraphQLSourceFromReScript(document);
    default:
      return null;
  }
}

export function extractSelectedOperation(
  languageId: string,
  document: string,
  selection: {
    line: number;
    character: number;
  }
): GraphQLSource | null {
  const sources = extractGraphQLSources(languageId, document);

  if (!sources || sources.length < 1) {
    return null;
  }

  let targetSource: GraphQLSource | null = null;

  if (sources[0].type === "FULL_DOCUMENT") {
    targetSource = sources[0];
  } else {
    // A tag must be focused
    for (let i = 0; i <= sources.length - 1; i += 1) {
      const t = sources[i];

      if (
        t.type === "TAG" &&
        selection.line >= t.start.line &&
        selection.line <= t.end.line
      ) {
        targetSource = t;
      }
    }
  }

  return targetSource;
}

export function getSelectedGraphQLOperation(
  doc: string,
  pos: { line: number; character: number }
): GraphQLSourceFromTag | null {
  const selectedOperation = extractSelectedOperation("rescript", doc, {
    line: pos.line,
    character: pos.character,
  });

  if (selectedOperation && selectedOperation.type === "TAG") {
    return selectedOperation;
  }

  return null;
}
