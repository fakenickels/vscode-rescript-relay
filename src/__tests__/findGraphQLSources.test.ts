import * as fs from "fs";
import * as path from "path";
import { extractGraphQLSourceFromReScript } from "../findGraphQLSources";

const fixture = fs.readFileSync(
  path.resolve(
    path.join(__dirname, "../", "testfixture", "graphqlSources.res")
  ),
  "utf8"
);

describe("findGraphQLSources", () => {
  it("finds ReScript sources", () => {
    expect(extractGraphQLSourceFromReScript(fixture)).toMatchSnapshot();
  });
});
