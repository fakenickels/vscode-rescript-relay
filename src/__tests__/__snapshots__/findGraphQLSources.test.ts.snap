// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findGraphQLSources finds ReScript sources 1`] = `
Array [
  Object {
    "content": "
  fragment SingleTodo_todoItem on TodoItem {
    id
    text
    completed
  }
",
    "end": Object {
      "character": 0,
      "line": 7,
    },
    "start": Object {
      "character": 3,
      "line": 1,
    },
    "type": "TAG",
  },
  Object {
    "content": "
  mutation SingleTodoDeleteMutation(
    $input: DeleteTodoItemInput!
    $connections: [ID!]!
  ) @raw_response_type {
    deleteTodoItem(input: $input) {
      deletedTodoItemId @deleteEdge(connections: $connections)
    }
  }
",
    "end": Object {
      "character": 0,
      "line": 20,
    },
    "start": Object {
      "character": 3,
      "line": 11,
    },
    "type": "TAG",
  },
  Object {
    "content": "
  mutation SingleTodoUpdateMutation($input: UpdateTodoItemInput!) {
    updateTodoItem(input: $input) {
      updatedTodoItem {
        id
        text
        completed
      }
    }
  }
",
    "end": Object {
      "character": 0,
      "line": 34,
    },
    "start": Object {
      "character": 3,
      "line": 24,
    },
    "type": "TAG",
  },
]
`;
